### ISSUE TITLE
Bud sends no X-Forwarded-For for a portion of the Chrome requests

### ISSUE THREAD

**Issue by phillipp** [indutny/bud#73]
We have a working bud setup now but we are seeing some connections where the first requests don't have an X-Forwarded-For header.

This very rare but breaks quite a lot of stuff when it happens.

Setup is a "normal" bud setup on 1.2.4 and a nginx-lua backend that reads and sets the X-Forwarded-For header into a SHM dictionary on the first request. To make sure we don't miss the first request we added a "seen" key that is set on the first request of the connection. We additionally dump all headers.

After a while I noticed that I can't reproduce with curl or Firefox, only in chrome. The headers confirmed that only chrome (like Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.132 Safari/537.36) is affected.
Accept-Encoding is always "gzip, deflate, sdch" (maybe helpful?), and we strip the Accept-Encoding at the nginx layer because of legacy system behaviour.

Reproduction is simple: just hit F5 a few times in chrome and on a few requests the IP is not set.

Any ideas?

**Comment by indutny** [comment]
Are you using single backend, or multiple ones? Are the backends loaded asynchronously using SNI?

**Comment by thomasba** [comment]
There is only one backend (a nginx server, so bud only handles the ssl stuff). The configuration:

```json
  "backend": [{
    "port": 10010,
    "host": "127.0.0.1",
    "keepalive": 3600,
    "x-forward": true
  }],
```

**Comment by indutny** [comment]
@phillipp may I ask you to try running bud a in with debug logging, and paste some results as a gist?

**Comment by indutny** [comment]
Mostly interested in the results at the moment of missing x-forwarded header.

**Comment by phillipp** [comment]
We tried some more (hitting F5 a lot) and could reproduce the same behaviour with Firefox, but not with Internet Explorer. Same with curl, could not reproduce it (maybe until now). I have a loop trying with curl over and over running for a while now, but the result looks always good.

Debug output of a request (hit enter in the stdout as a "mark" and F5 in the browser until behaviour shows up) listed in the gist:
https://gist.github.com/phillipp/a466d13c83902acef9be

**Comment by phillipp** [comment]
Okay, this may sound crazy but: could this be a result of a race condition in checking which request is the first one of a connection? I think this *could* be the case because:

- I can't reproduce this on LAN, only on WAN (>40ms ping)
- The more assets are loaded on the page, the higher the problem rate

**Comment by krtschmr** [comment]
i can confirm the bug @phillipp reported. tested together with him and ran into same problems

**Comment by indutny** [comment]
@phillipp this is very interesting idea! If there is nothing secret happening on the machine - may I ask you to run a `tcpdump -w dump.pcap` on it, and send me the results? It should be pretty straightforward to confirm or disprove this theory.

**Comment by phillipp** [comment]
There is a lot of customer traffic running on this machine but we can tcpdump only the relevant ip/port combination. Would you need the backend traffic, too? Or just the frontend traffic?

**Comment by indutny** [comment]
Actually, just a backend traffic at the time of failure. Thanks!

**Comment by phillipp** [comment]
Sorry, it took a bit longer to generate the file: https://storage.googleapis.com/trafficplexgmbh/bud-out.cap

I captured the background traffic from bud. In the last ~8 requests there should be two requests on a new connection without a X-Forwarded-For header.

You can finde the X-Forwarded-For header in the PHP info-Output at REMOTE_ADDR in the "Apache 2" section.

**Comment by indutny** [comment]
@phillipp from the cap output that you sent me - it looks like the first request on every TCP connection to backend has the X-Forwarded-For header. Subsequent requests on the same connection do not have it, but this is how it works in bud.

Am I missing something?

**Comment by phillipp** [comment]
I only wee what nginx is giving me: new connection ID, no header. Maybe that's a bug in nginx?

**Comment by indutny** [comment]
@phillipp certainly something about either nginx or lua backend that you are using.

**Comment by phillipp** [comment]
Ugh, that's strange. Sorry to waste your time!

**Comment by indutny** [comment]
No worries, @phillipp . I'm sad that I can't help much with it

**Comment by phillipp** [comment]
We added some debug stuff to nginx and came up with the following theory:

a) Nginx buffers all incoming requests and only starts processing the request if it is fully received. That is expected behaviour.
b) The connection ID is set on the `accept` event.
c) When the full request is read, the requests counter of the connection is increased (starting with 1 for the first request).
d) The X-Forwarded-For header is processed and added to the SHM dict after the full request is read. The SHM dictionary key is the connection ID. But for requests where the X-Forwarded-For header is missing, the request counter is > 1, so there were requests processed on the connection that buffered "in less time" and were processed before the request with the X-Forwarded-For header on the same connection.

So basically: it sometimes happens that the first requests bud sends (with the X-Forwarded-For header) is not the first request nginx processes on the connection, but the second, third or even later one because all requests are fully buffered. 

Because nginx does not support stream processing for the requests, would it theoretically be possible to add X-Forwarded-For headers for all requests in bud?

**Comment by indutny** [comment]
@phillipp hm... I did consider that previously, but it would imply double-http parsing and code complexity in bud.

Is there any chance you could just use `proxyline` and parse it using some lua pluging in nginx before accepting *any* requests?

**Comment by phillipp** [comment]
Ha, that lead to the right module. I've added an example config to the docs: https://github.com/indutny/bud/pull/75
